open-company/open-company-auth
# persist bot tokens in S3

[Trello card](https://trello.com/c/QEAeBkKE/428-org-bot-id-storage-in-auth-include-in-jwtokens) — depends on https://github.com/belucid/open-company-infrastructure/pull/18

This adds a small layer of persistence to the Auth service to store secrets we need might need to trigger bot conversations. It stores these secrets in S3 for now. As needs evolve we can re-evaluate other options. (Previous discussions suggested Redis & RethinkDB, I'm now thinking DynamoDB via [faraday-atom](https://github.com/danstone/faraday-atom/) might also be a good option.)

The `bot-integration` [PR to API](https://github.com/open-company/open-company-api/pull/43) has been updated to use the presence of bot tokens as requirement instead of checking that the user is an owner of the team: https://github.com/open-company/open-company-api/pull/43/commits/a513a8e2cd98d50eeec9bbf85d16f3719af44929 

The token-writing is synchronous so authentication may take a few hundred ms extra. This should only be the case if a bot token is present and it's different to the one already stored in the database so a minor portion auth-flows. 

**To Test:**
- sign out
- sign in via Slack: **make sure the scopes include `bot`**
- check the `store` file in the S3 bucket
  - it should contain your Slack org-id
  - and the bot-id and the bot-secret

You can also try the same with the orgs/users provided in the infrastructure repo. Once an owner has signed in all tokens issued afterwards should contain the bot secrets (and these should be stored in the `store`/`store-production` files in the S3 bucket).
